sklearn/tests/metadata_routing_common.py

```diff
@@ -147,12 +148,11 @@ def partial_fit(self, X, y, sample_weight="default", metadata="default"):
         return self
 
     def fit(self, X, y, sample_weight="default", metadata="default"):
-        if self.registry is not None:
-            self.registry.append(self)
-
         record_metadata_not_default(
             self, "fit", sample_weight=sample_weight, metadata=metadata
         )
+        if self.registry is not None and self._records.get("fit", {}):
+            self.registry.append(self)
```

Comment:
I have also changed the conditions for when a registry is build in `ConsumingClassifier` and `ConsumingRegressor`, so that the test can accommodate for passing a metadata to one consumer, while not to the other.

More generally, I'm still wondering about the registry. Is this a common helper in testing? Why is it build in the mocking estimators and not in the tests? It feels like we're testing something while also giving part of the solution away.

Reply:
Concerning the CI failure at `assert estimator in estimator.registry` from `test_metadata_routing`:

Okay, having a mocking estimator put itself on the registry **on condition** that this type of metadata is routed to  it, leads to that error in `test_estimator_puts_self_in_registry`. 

Now I'm not sure what is more valuable: the stability of knowing each estimator appears in the registry or the flexibility for testing?

Reply:
If you have different sub-estimators and they get different metadata, then you should pass them two different instances of the `_Registry`, and move this condition back to where it was.

The mocking module is also used in tests, I don't understand why it's not the right place for it.

Reply:
Yes, that now makes sense. 

I've now build different registries for different sub-estimators (in `test_metadata_routing_for_voting_estimators`) they take different metadata.